core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -3143,63 +3140,69 @@ private void mountInternal(RpcContext rpcContext, LockedInodePath inodePath, All
       throw new InvalidPathException(
           ExceptionMessage.MOUNT_POINT_ALREADY_EXISTS.getMessage(inodePath.getUri()));
     }
+    // validate the Mount operation first
+    mountPathValidation(inodePath, ufsPath);
     long mountId = IdUtils.createMountId();
-    mountInternal(rpcContext, inodePath, ufsPath, mountId, context);
+    // get UfsManager prepared
+    prepareUfsForMount(mountId, ufsPath, context);
     boolean loadMetadataSucceeded = false;
     try {
       // This will create the directory at alluxioPath
-      InodeSyncStream.loadDirectoryMetadata(rpcContext,
+      InodeSyncStream.loadMountPointDirectoryMetadata(rpcContext,
           inodePath,
           LoadMetadataContext.mergeFrom(
               LoadMetadataPOptions.newBuilder().setCreateAncestors(false)),
-          mMountTable,
+          context.getOptions().getShared(), ufsPath, mUfsManager.get(mountId),
           this);
       loadMetadataSucceeded = true;
     } finally {
-      if (!loadMetadataSucceeded) {
-        mMountTable.delete(rpcContext, inodePath.getUri(), true);
+      if (loadMetadataSucceeded) {
+        // As we have verified the mount operation by calling MountTable.verifyMount, there won't
+        // be any error thrown when doing MountTable.add
+        mMountTable.add(rpcContext, inodePath.getUri(), ufsPath, mountId,
+            context.getOptions().build());
+      } else {
+        mUfsManager.removeMount(mountId);
       }
     }
   }
 
+  /**
+   * Get UfsManager prepared before add/update MountTable.
+   * @param mountId the mountId of this new mount operation
+   * @param ufsPath the ufs path to be mounted
+   * @param context the mounting context
+   * @throws IOException can be thrown when the ufsPath doesn't exist
+   */
+  private void prepareUfsForMount(long mountId, AlluxioURI ufsPath, MountContext context)
```

Comment:
This wrapper does very little because half of the real logic is in yet another wrapper. Can you somehow merge or refactor these wrappers?